### ISSUE TITLE
Potential pull request for pre-allocated buffer read?

### ISSUE THREAD

**Issue by ckemere** [larsimmisch/pyalsaaudio#152]
I use read() in a tight loop where I write microphone data to a file using python-soundfile. I've found it helpful to create a version of the read function that takes a preallocated numpy buffer, so I can write code like this:
```
import soundfile
import numpy as np

# Open alsa device
adevice = alsaaudio.PCM(alsaaudio.PCM_CAPTURE, alsaaudio.PCM_NORMAL, device=device)

self.adevice.setchannels(num_channels) 
self.adevice.setrate(fs)
scale_factor = 4 # system fixed number of buffers
if dtype == 'int16':
    adevice.setformat(alsaaudio.PCM_FORMAT_S16_LE)
    scale_factor = scale_factor*2
else:
    raise(ValueError("dtypes other than 'int16' not currently supported."))

adevice.setperiodsize(self.buffer_size)
in_buf = np.zeros((buffer_size*scale_factor, channels), dtype=dtype, order='C')

with soundfile.SoundFile(soundfilename, 'w', fs, channels, 'PCM_16') as soundfile:
    while True:
        nsamp = adevice.read_into(in_buf)
        self.soundfile.write(in_buf[:nsamp,:])
        if nsamp < buffer_size:
            print('ALSA Read buffer underrun.')
```

The two caveats about the proposed`read_into()` are (1) dependency on numpy and (2)  I only have working Python 3 code.

**Comment by larsimmisch** [comment]
Only Python 3 would be fine, but a default compile-time dependency on numpy would be problematic. 

Do you see a way to disentangle this? Would it be possible to make this optional at compile-time?

**Comment by ossilator** [comment]
i wonder whether a hard dep on numpy would be really a problem. what actual uses outside the scientific community (which uses numpy throughout) does pyaa have?

conversely, i somewhat doubt that buffer pre-alloc actually buys you a lot in this case, as even the tightest loop is (sample-)rate-limited.

anyway, one could do it without numpy while being numpy-compatible by using the buffer protocol directly. that would make the usage somewhat uglier, though.

**Comment by ckemere** [comment]
Thanks for commenting back so quickly! I agree that the memory allocation is probably not a huge drain on resource. I think I can also potentially trigger garbage collection on the input buffer. My naive implementation just using read() currently also involves copying the data from the input buffer into a numpy array and reshaping it, along with error handling if it's the wrong size. So I actually end up effectively doubling the memory allocation/garbage collection load. Any thoughts on that?

I looked briefly at the buffer protocol documentation, and haven't fully understood it yet. I'm not really a python expert, and so there are things that I don't fully understand how to do. It seems like I should be able to effectively cast the `bytes` returned by read() into the memory area of a numpy array, but I can't figure out how to do that...

**Comment by ckemere** [comment]
Ah, actually, reading more, it seems that 1np.frombuffer` creates a view rather than a copy, so perhaps it's less of an issue.

**Comment by ossilator** [comment]
yes, my own code uses frombuffer() and reshape()'s it according to sample size and channel count.

**Comment by ckemere** [comment]
`np.array()` will copy, though, right?

**Comment by RonaldAJ** [comment]
> `np.array()` will copy, though, right?

I think so, but you can perhaps use frombuffer there. I would have to look into it to be sure.

**Comment by miccoli** [comment]
A little late into this thread but let me add some comments.

A `readinto` method, modeleld onto the semantics of [`RawIOBase.readinto`](https://docs.python.org/3/library/io.html#io.RawIOBase.readinto) has it merits but this is completely othogonal to the conversion of raw audio data to numpy arrays, which can be easily obtained with zero copy (and without using `struct.unpack`).

Let me spell out explicilty how `frombuffer` and `reshape` can be used.

```python
>>> rawdata = b'\x00\x00\x01\x00\x02\x00\x03\x00\x04\x00\x05\x00\x06\x00\x07\x00'
>>> data = np.frombuffer(rawdata, np.dtype('<u2'))
>>> data
array([0, 1, 2, 3, 4, 5, 6, 7], dtype=uint16)
>>> data.base is rawdata
True
>>> data2 = data.reshape(-1, 2)
>>> data2
array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7]], dtype=uint16)
>>> data2.base.base is rawdata
True
```
In this case `data`/`data2` are read-only views into the bytes buffer `buf`. It is read-only because bytes are immutable in python. If you need a read-write view, then the following pattern can be useful:
```python
>>> rawdata = bytearray(b'\x00\x00\x01\x00\x02\x00\x03\x00\x04\x00\x05\x00\x06\x00\x07\x00')
>>> data = np.frombuffer(rawdata, np.dtype('<u2'))
>>> data += 256
>>> rawdata
bytearray(b'\x00\x01\x01\x01\x02\x01\x03\x01\x04\x01\x05\x01\x06\x01\x07\x01')
```

The numpy datatype system can cope zero-copy also with (non native) BigEndian rawdata:
```
>>> rawdata = bytearray(b'\x00\x00\x01\x00\x02\x00\x03\x00\x04\x00\x05\x00\x06\x00\x07\x00')
>>> data = np.frombuffer(rawdata, np.dtype('>u2'))
>>> data
array([   0,  256,  512,  768, 1024, 1280, 1536, 1792], dtype='>u2')
>>> data.base.obj is rawdata
True
```

Once  an optimized `readinto` is implemented for generic buffers, it can be used also for filling numpy ndarrays:
```
npbuf = np.empty((1024,), dtype="<u2")
adevice.read_into(npbuf)
```
since ndarray implements the buffer protocol:
```
>>> rawdata = bytearray(b'\x00\x00\x01\x00\x02\x00\x03\x00\x04\x00\x05\x00\x06\x00\x07\x00')
>>> f = io.BytesIO(rawdata)
>>> npbuf = np.zeros((8,), dtype="<u2")
>>> f.readinto(npbuf)
16
>>> npbuf
array([0, 1, 2, 3, 4, 5, 6, 7], dtype=uint16)
>>> npbuf = np.zeros((8,), dtype=">u2")
>>> f.seek(0)
0
>>> f.readinto(npbuf)
16
>>> npbuf
array([   0,  256,  512,  768, 1024, 1280, 1536, 1792], dtype='>u2')
